File: neetcode-150/Two_pointers/Valid_palindrome/solution.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        # Convert the string to lowercase for case-insensitive comparison
        s = s.lower()

        # Set up two pointers: one at the start and one at the end
        left, right = 0, len(s) - 1

        while left < right:
            # Skip non-alphanumeric characters from the left
            while left < right and not s[left].isalnum():
                left += 1
            
            # Skip non-alphanumeric characters from the right
            while left < right and not s[right].isalnum():
                right -= 1
            
            # Compare characters; return False if they don't match
            if s[left] != s[right]:
                return False

            # Move both pointers inward
            left += 1
            right -= 1

        # All characters matched; it's a palindrome
        return True

File: neetcode-150/Two_pointers/Valid_palindrome/solution.js

function isPalindrome(s){
    let sol = new Solution()

    return sol.validPalindrome(s)
}

class Solution{
    validPalindrome(s){
        // 1. convert the string to lowercase
        let s_lower = s.toLowerCase();

        // 2. set up two pointers : one at start and one at the end
        let left = 0, right = s_lower.length - 1;

        // 3. while loops whenever left < right
        while(left < right){
             // 4. skip the non alphanumeric chars from left
             while(left < right && !/^[a-z0-9]+$/i.test(s_lower[left])){
                left++
             }

            // 5. skip the non alphanumeric chars right
             while(left < right && !/^[a-z0-9]+$/i.test(s_lower[right])){
                right--
             } 

             // 6. compare left end and right end chars return false if not equal
             if(s_lower[left] !== s_lower[right]){
                return false
             }
             left++
             right--
        }
        return true
    }
}